csharp/src/Apache.Arrow/Arrays/Array.cs

```diff
@@ -59,5 +59,22 @@ internal static void Accept<T>(T array, IArrowArrayVisitor visitor)
                     break;
             }
         }
+
+        public Array Slice(int offset, int length)
+        {
+            if (offset > Length)
+            {
+                throw new ArgumentException($"Offset {offset} cannot be greater than Length {Length} for Array.Slice");
+            }
+
+            length = Math.Min(Data.Length - offset, length);
```

Comment:
Should this just throw if you asked for too long of a length?

Reply:
Throwing an ArgumentOutOfRangeException is how `Span<T>` works in .NET (and honestly any other API I've seen in the BCL). If a caller passes in a value that is too large, it is typically a mistake, and we should tell them.

---
In reply to: [262338881](https://github.com/apache/arrow/pull/3806#discussion_r262338881) [](ancestors = 262338881)

Reply:
The C++ version doesn't throw and is doing a min call. I don't mind throwing here in C#, but is that what we want? I don't *think* diverging from C++ behavior here is a significant concern?

Reply:
Ok, I’m fine with mimicking the C++ behavior.

Reply:
I don't think there's a need to mimic the C++ library exactly, whatever seems most idiomatic and reasonable in C# (w.r.t. error handling) -- in some cases in C++ we are relying on debug-time checks for programmer errors